assembly/src/assembler/instruction/crypto_ops.rs

```diff
@@ -145,27 +145,41 @@ pub(super) fn mtree_set(span: &mut SpanBuilder) -> Result<Option<CodeBlock>, Ass
     // stack: [d, i, R_old, V_new, ...]
 
     // stack: [V_old, R_new, ...] (29 cycles)
-    update_mtree(span, true)
+    update_mtree(span)
 }
 
-/// Appends the MRUPDATE op with a parameter of "true" and stack manipulations to the span block as
-/// required to copy a Merkle tree with root R and update the node in the copied tree at depth d
-/// and index i to value V. The stack is expected to be arranged as follows (from the top):
-/// - depth of the node, 1 element; this is expected to be the depth of the Merkle tree
-/// - index of the node, 1 element
-/// - current root of the tree, 4 elements
-/// - new value of the node, 4 element
+/// Injects [AdviceInjector::MerkleMerge] into the advice provider. The stack is expected to be
+/// arranged as follows (from the top):
+/// - root of the right tree, 4 elements
+/// - root of the left tree, 4 elements
 ///
-/// After the operations are executed, the stack will be arranged as follows:
-/// - old value of the node, 4 elements
-/// - new root of the tree after the update, 4 elements
+/// The operation will merge the Merkle trees with the provided roots, producing a new merged root
+/// with incremented depth. After the operations are executed, the stack will be arranged as
+/// follows:
+/// - merged root, 4 elements
 ///
-/// This operation takes 29 VM cycles.
-pub(super) fn mtree_cwm(span: &mut SpanBuilder) -> Result<Option<CodeBlock>, AssemblyError> {
-    // stack: [d, i, R_old, V_new, ...]
+/// This operation will fail if any of the input roots doesn't exist as Merkle tree in the advice
+/// provider.
+///
+/// This operation takes 12 VM cycles.
+pub(super) fn mtree_merge(span: &mut SpanBuilder) -> Result<Option<CodeBlock>, AssemblyError> {
+    // stack input:  [R_rhs, R_lhs, ...]
+    // stack output: [R_merged, ...]
 
-    // stack: [V_old, R_new, ...] (29 cycles)
-    update_mtree(span, true)
+    // uses the top 2 words from the operand stack to perform a Merkle merge and expect the merged
+    // node to be pushed to the advice stack
+    span.push_decorator(Decorator::Advice(AdviceInjector::MerkleMerge));
+
+    #[rustfmt::skip]
+    let ops = [
+        // drops the two merged words (8 elements) from the operand stack
+        Drop, Drop, Drop, Drop,
+        Drop, Drop, Drop, Drop,
+
+        // reads the merged root (4 elements) from the advice stack
+        Read, Read, Read, Read,
```

Comment:
We can't quite do it this way since the advice provider is not trusted and there is no guarantee that the values we'd read from it would actually represent `hash(left, right)`. But we can do it slightly differently:

- First, the `MerkleMerge` decorator doesn't need to put anything on the advice stack. The only thing it needs to do is update the `MerkleStore` (as it does now).
- Second, instead of dropping the values here and then reading from the advice stack, we can execute the equivalent of `hmerge` instruction (or maybe we can just invoke the [hmerge](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/assembler/instruction/crypto_ops.rs#L73-L97) function directly here).

So, what this does is that `hmerge` will compute the hash within the VM, and `MerkleMerge` decorator will synchronize the Merkle store to have a new tree with the same root.

Reply:
Done! I think it made more sense to just call `hmerge` there. We run the risk of having outdated cycles documentation if that changes for hmerge, but then we guarantee compatibility.